htmresearch/algorithms/column_pooler.py

```diff
@@ -174,16 +192,127 @@ def compute(self, feedforwardInput=(), lateralInputs=(),
+
+    @param  lateralInputs (list of sequences)
+            For each lateral layer, a list of sorted indices of active lateral
+            input bits
+
+    @param  feedforwardGrowthCandidates (sequence or None)
+            Sorted indices of feedforward input bits that the active cells may
+            grow new synapses to.  This is assumed to be the predicted active
+            cells of the input layer.
+    """
+    prevActiveCells = self.activeCells
+    cellsToLearn = set(self.activeCells)
+
+    # If there are not enough previously active cells, select random subset of
+    # cells to learn on.  If there were only some active cells, we assume that
+    # we are learning an object related to what is represented by those cells;
+    # if there were none, we are learning a new object.
+    # This case is the only way different object representations are created.
+    if len(cellsToLearn) < self.sdrSize:
+      if len(cellsToLearn) < self.sdrSize*(1 - self.learningTolerance):
+        cellsToLearn = set()
+      numToAdd = self.sdrSize - len(cellsToLearn)
+      newCells = _sampleRange(self._random,
+                              0, self.numberOfCells(),
+                              step=1, k=numToAdd)
+      cellsToLearn |= set(newCells)
+
+      cellsToLearn = numpy.asarray(list(cellsToLearn))
+      cellsToLearn.sort()
+      self.activeCells = numpy.union1d(self.activeCells, cellsToLearn)
+      self.activeCells.sort()
+
+      # Form initial internal distal links for the new SDR.  This is crucial in
+      # online learning.  Without it, the SDR may fail to persist.
+      self._learn(self.internalDistalPermanences, self._random,
+                  self.activeCells, self.activeCells, self.activeCells,
+                  self.sampleSizeDistal, self.initialDistalPermanence,
+                  self.synPermDistalInc, self.synPermDistalDec,
+                  self.connectedPermanenceDistal)
+
+    else:
+      cellsToLearn = numpy.asarray(list(cellsToLearn))
+      cellsToLearn.sort()
+
+    # If we have a union of cells active, don't learn.
+    if len(cellsToLearn) > self.sdrSize*(1+self.learningTolerance):
+      return
+
+    cellsToLearn.sort()
+    self.activeCells = numpy.union1d(self.activeCells, cellsToLearn)
+    self.activeCells.sort()
+
+    # Finally, now that we have decided which cells we should be learning on, do
+    # the actual learning.
+    if len(feedforwardInput) > 0:
+      self._learn(self.proximalPermanences, self._random,
+                  cellsToLearn, feedforwardInput,
+                  feedforwardGrowthCandidates, self.sampleSizeProximal,
+                  self.initialProximalPermanence, self.synPermProximalInc,
+                  self.synPermProximalDec, self.connectedPermanenceProximal)
+
+      # External distal learning
+      # We should do this no matter what, since other columns might still be
+      # learning useful things even if we're not.
```

Comment:
Is this comment saying that we should do this even when there's no feedforward input? (That's not what the code is doing.)